#721: Add matcher for TeeInput class

proshin-roman opened the pull request:
It is for issue #721: adds a new matcher that checks result of TeeInput class: compares both TeeInput's results (direct and copied) to the expected value. As well this new matcher has been applied to existing tests of TeeInput class.

0crat commented:
Job [#731](https://github.com/yegor256/cactoos/issues/731) is now in scope, role is `REV`

<!-- https://www.0crat.com/footprint/C63314D6Z/691534699, version: 1.0-SNAPSHOT, hash: 4679fa60 -->

0crat commented:
This pull request [#731](https://github.com/yegor256/cactoos/issues/731) is assigned to @neonailol[/z](https://www.0crat.com/u/neonailol), here is [why](https://www.0crat.com/footprint/C63314D6Z/691536760). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

<!-- https://www.0crat.com/footprint/C63314D6Z/691536767, version: 1.0-SNAPSHOT, hash: 4679fa60 -->

proshin-roman commented:
@neonailol ping :)

neonailol commented on src/main/java/org/cactoos/matchers/TeeInputHasResult.java:
```diff
@@ -0,0 +1,136 @@
+     */
+    public TeeInputHasResult(
+        final Text expected,
+        final Text copied) {
```
formatting is not consistent

neonailol commented on src/main/java/org/cactoos/matchers/TeeInputHasResult.java:
```diff
@@ -0,0 +1,136 @@
+     */
+    public TeeInputHasResult(
+        final String expected,
+        final Text copied) {
```
formatting is not consistent

neonailol commented on src/main/java/org/cactoos/matchers/TeeInputHasResult.java:
```diff
@@ -0,0 +1,136 @@
+    public boolean matchesSafely(final TeeInput item) {
+        this.actual = new ComparableText(new TextOf(item));
+        return
+            this.expected.compareTo(this.actual) == 0
```
Please use `org.cactoos.scalar.Equals` here

neonailol commented on src/main/java/org/cactoos/matchers/TeeInputHasResult.java:
```diff
@@ -0,0 +1,136 @@
+                new FormattedText(
+                    TeeInputHasResult.DESCRIPTION,
+                    new UncheckedText(this.expected).asString(),
+                    new UncheckedText(this.expected).asString()
```
Duplicated line `new UncheckedText(this.expected).asString()`

proshin-roman commented:
@neonailol Thanks for your comments - they were useful. I've fixed all of them except a comment about inconsistent formatting: I've changed another ctor where the closing bracket `)` is on a new line. Please review the pull request again.

neonailol commented:
@proshin-roman thanks looks good now

neonailol commented:
@rultor good to merge

rultor commented:
>  @rultor good to merge

@neonailol Thanks for your request. @yegor256 Please confirm this.

proshin-roman commented:
@yegor256 ping

neonailol commented:
@yegor256 Please confirm this.

yegor256 commented:
@proshin-roman to be honest, I don't like this at all. Why do we need this matcher? What's wrong with existing `assertThat` two times? It looks like something very artificial and super-specific for one particular use case.

proshin-roman commented:
@yegor256 well, TeeInput class has a lot of ctors (about 50) and all of them should be covered by tests (there is a ticket for that). It means double-assert will be copy-pasted for all these test cases. Another problem of double-assert: you have to trigger them in exact order otherwise they fail. This new matcher allows to avoid both of these problems. Finally, it's a goal of the ticket - you can decline this feature at all as the Arch of the project. And then I will have a reason to use double-assert in my other pull request.

0crat commented:
@neonailol[/z](https://www.0crat.com/u/neonailol) this job was assigned to you 8 days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

<!-- https://www.0crat.com/footprint/C63314D6Z/771534796, version: 1.0-SNAPSHOT, hash: ec360c06 -->

yegor256 commented:
@rultor merge

rultor commented:
> @rultor merge

@yegor256 OK, I'll try to merge now. You can check the progress of the merge [here](http://www.rultor.com/t/14123-374523509)

proshin-roman commented:
@yegor256 damn, the pull request is going to be merged but the ticket is already resigned from me. I would be very happy if you assign it back to me, please :)

0crat commented:
@elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova) please review this job completed by @neonailol[/z](https://www.0crat.com/u/neonailol), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

<!-- https://www.0crat.com/footprint/C63314D6Z/790931359, version: 1.0-SNAPSHOT, hash: faa47291 -->

0crat commented:
The job [#731](https://github.com/yegor256/cactoos/issues/731) is now out of scope

<!-- https://www.0crat.com/footprint/C63314D6Z/790931369, version: 1.0-SNAPSHOT, hash: faa47291 -->

0crat commented:
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

<!-- https://www.0crat.com/footprint/C63314D6Z/790932374, version: 1.0-SNAPSHOT, hash: faa47291 -->

elenavolokhova commented:
@neonailol According to our Policy:
>The code reviewer found at least three problems in the code

Only two issues were fixed and actually improved the code quality in this case. 
Please, confirm that you will try to find at least three problems next time.

neonailol commented:
@elenavolokhova i confirm

elenavolokhova commented:
@0crat quality acceptable

0crat commented:
Order was finished, quality is "acceptable": +15 points just awarded to @neonailol[/z](https://www.0crat.com/u/neonailol)

<!-- https://www.0crat.com/footprint/C63314D6Z/800538814, version: 1.0-SNAPSHOT, hash: faa47291 -->

0crat commented:
Quality review completed: +8 points just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

<!-- https://www.0crat.com/footprint/C63314D6Z/800538816, version: 1.0-SNAPSHOT, hash: faa47291 -->
